ui/src/app/zones/views/ZonesListForm/ZonesListForm.test.tsx

```diff
@@ -0,0 +1,93 @@
+import { mount } from "enzyme";
+import { act } from "react-dom/test-utils";
+import { Provider } from "react-redux";
+import configureStore from "redux-mock-store";
+
+import ZonesListHeader from "../ZonesListHeader/ZonesListHeader";
+
+import type { RootState } from "app/store/root/types";
+import {
+  // zone as zoneFactory,
+  // zoneState as zoneStateFactory,
+  rootState as rootStateFactory,
+} from "testing/factories";
+
+const mockStore = configureStore();
+let state: RootState;
+
+describe("ZonesListForm", () => {
+  beforeEach(() => {
+    state = rootStateFactory();
+  });
+
+  it("displays the form when Add AZ is clicked", () => {
+    const store = mockStore(state);
+    const wrapper = mount(
+      <Provider store={store}>
+        <ZonesListHeader />
+      </Provider>
+    );
+    expect(wrapper.find("ZonesListForm").exists()).toBe(false);
+
+    wrapper.find("button[data-test='add-zone']").simulate("click");
+
+    expect(wrapper.find("ZonesListForm").exists()).toBe(true);
+  });
+
+  it("calls actions.create on save click", () => {
+    const store = mockStore(state);
+    const wrapper = mount(
+      <Provider store={store}>
+        <ZonesListHeader />
+      </Provider>
+    );
+    wrapper.find("button[data-test='add-zone']").simulate("click");
+
+    act(() =>
+      wrapper.find("Formik").props().onSubmit({
+        name: "test-zone",
+      })
+    );
+
+    expect(
+      store.getActions().find((action) => action.type === "zone/create")
+    ).toStrictEqual({
+      type: "zone/create",
+      meta: {
+        method: "create",
+        model: "zone",
+      },
+      payload: {
+        params: {
+          name: "test-zone",
+        },
+      },
+    });
+  });
+
+  // it("checks if the form closes on successful submit", () => {
```

Comment:
@hatched Oh, thanks for the comment. We talked a bit about that with Scott and Beth, and while I agree that these kind of tests (interact as user would) don't really fit into unit testing with jest, mocked stores and actions. This feels more like a test you would write in cypress.

Jest unit tests for components seem to be more like: in case of such store state, that's what should render, and when user clicks, that's the redux action we expect. So yes, it depends more on implementation (redux store and actions), but only from component perspective.

So in case of a test "form gets closed after successful subimit" - is it something that should be unit tested with jest, or integration tested with cypress? Because creating such integration test in jest seems quite complex and requires a lot of knowlege about internal implementation anyway (how to mock the store data, what actions are dispatched, how to mock formik events, etc).

Reply:
In my mind there are three generally accepted styles of testing, unit, API, and integration. In the React world unit testing is falling out of favor because the community (rightly so) believes that you should be testing the component as a user would be using it. As your tests become further from how they operate in reality the less confidence you can have in your test.

That leaves us with API and integration testing and the fuzzy distinction between the two. The setup of [Enzyme](https://enzymejs.github.io/enzyme/) and [React Testing Library](https://testing-library.com) are designed around allowing you to simulate rendering a component and interacting with it like it's in a real DOM. They both use [jsdom](https://github.com/jsdom/jsdom) under the hood.

To help differentiate API and integration testing, what does API testing even mean in relation to a React component? Well most components have three different API surfaces, the dom, props, and data (think redux). It's these surfaces that the various types of users will interact with, the developer will interact with the props and data surface and the user will interact with the dom. Using the aforementioned libraries we can test all three of these API surfaces with a high level of confidence. We can wrap the component in the necessary mocks for the data API, we can render the component with the appropriate props, and using jsdom (or the libraries) we can interact with the rendered output as well.

Swinging back to the actual test in question now. We want to ensure that the component behaves in a certain way given a specific operating context. We aren't concerned about how it'll operate in the application, only how the component itself functions when it's API surface area is exercised in various ways. So to this end we can do things like setup a mock data layer to ensure that the component sends the correct data out, we can modify the props so that we can test it's various rendering states, and we can simulate DOM interactions to ensure that it behaves correctly.

And that's the difference between integration tests. That's testing how the component will operate when integrated within the application as a whole. In this case, yes you will still be testing that the props are being passed in correctly, that the dom interactions work as expected and that the data gets updated correctly but you'll be doing so in context of the application. So you can see if that submit button is still working or if some other component inadvertently covers it up. Or if it can handle the data in the store that some other component has put there.

Wow that ended up being a long one :D I hope that helps clear up where I draw the line between the type of testing patterns.

Reply:
I wrote a blog post that touches on this from a different angle that you might find interesting [Testing your user contract](https://fromanegg.com/post/2020/01/01/testing-your-user-contract/)

Reply:
@hatched Let's not let perfect get in the way of good, especially in a hackathon-style sprint. I agree with your philosophy, but I think @bartaz suggestion that this falls more in line with a cypress integration test is apt. [We've already done this for adding a machine](https://github.com/canonical-web-and-design/maas-ui/blob/master/integration/cypress/integration/machines/add.spec.ts). You'll notice it's 6 lines of code, but it at least gives us some confidence that a real MAAS and real user input leads to the correct result.

I would consider a test in `FormikForm` that asserts that the `onSuccess` function is called properly to be enough for now, with cypress tests to come later.